README.rst
nens-graph
==========================================

Nens-graph is a separate library of custom matplotlib graphs for use
in the lizard-framework.

The graph layout should be controlled in the library, and only the data
should be added in the calling Django application. Common classes useful
for other graphs go in common.py, stuff changing from graph to graph should
be in eachs graphs own class.


CHANGES.rst
Changelog of nens-graph
===================================================


0.14 (unreleased)
-----------------

- Nothing changed yet.


0.13 (2012-06-21)
-----------------

- Added DateGridGraph.timeseries_as_list for displaying timeseries as
  tables.


0.12 (2012-05-23)
-----------------

- Added NensGraph.render function with options for: bmp, emf, eps,
  pdf, ps, raw, rgb, rgba, svg, svgz, png (default). Note that there
  is no .jpg or .gif option.


0.11 (2012-05-22)
-----------------

- Added option 'reversed_legend_items' in DataGridGraph.legend.


0.10 (2012-05-22)
-----------------

- Added option 'remove_duplicates' for DataGridGraph.legend.


0.9 (2012-04-23)
----------------

- Legends now have 100 pixels instead of 90 pixels.


0.8 (2012-03-26)
----------------

- Added function dates_values_comments. The function dates_values is
  now a wrapper for dates_values_comments.


0.7 (2012-02-15)
----------------

- Made matplotlib markers possible in GraphItem line.


0.6.1 (2012-02-13)
------------------

- Improved margins around legend when legend-location is 7.


0.6 (2012-02-09)
----------------

- Added timeseries_csv output for DateGridGraph.


0.5.2 (2012-02-09)
------------------

- Changed MARGIN_LEFT from 96 to 104.


0.5.1 (2012-02-08)
------------------

- Fixed buildout config.

- Fixed bug using stacked bars.


0.5 (2012-02-07)
----------------

- Moved DateGridGraph from lizard-graph to here.


0.4.5 (2011-11-21)
------------------

- Added MANIFEST.in for making packages.


0.4.4 (2011-11-21)
------------------

- Added new scales: RestrictToMonthScale, MercatorLatitudeScale.

- Added kwarg response to NensGraph.png_response to make it more flexible.


0.4.3 (2011-10-06)
------------------

- Used latest layout system for opendap graph, inspired by rainappgraph

- Added dedicated method for getting bbox of ticklabels, since the general way
  didn't work.

- Made the truncation length of legend labels a little larger


0.4.2 (2011-10-06)
------------------

- Put legend back in place


0.4.1 (2011-10-05)
------------------

- Adjusted Opendap graph so that it does not define locator and formatter if the
  restrict_to_month keyword is present.


0.4 (2011-09-30)
----------------

- Nothing changed yet.


0.3.1 (2011-09-19)
------------------

- Added use of tz kwarg in Rainapp graph


0.3 (2011-09-07)
----------------

- Changed RainappGraph.legend() so that it adds a legend object to the graph.

- Moved handling of figure width and height to common.NensGraph


0.2 (2011-09-01)
----------------

- Cleaned up rainapp graph. It is now only intended for a barchart of
  precipitation data.

- Implemented a custom ylim calculation and setting for barcharts.

- Set lower ylim to -1% and ylim max to at least 1 mm in rainapp.


0.1 (2011-08-30)
----------------

- Changed signature of NensGraph.on_draw()

- Added RainappGraph

- Added pixel-to-figure-coordinates methods for NensGraph class

- Added methods to find out width and height of specific plot objects at drawing
  time


0.0.1 (2011-05-18)
------------------

- Initial library skeleton created by nensskel.

- Copied orinal lizard-map.adapter.graph to basic.

- Added common graph class and deltaportaal-specific river graph


nens_graph/river.py
from math import pi
from random import choice
from random import seed
from numpy import arange
from numpy import array
from numpy import concatenate
from logging import getLogger
from nens_graph.common import NensGraph

from matplotlib import cm
from matplotlib.collections import RegularPolyCollection
from matplotlib.collections import CircleCollection
from matplotlib.lines import Line2D
from matplotlib.text import Text

logger = getLogger(__name__)


class RiverGraph(NensGraph):
    """Class for matplotlib river graphs."""

    def __init__(self,
                 start_km=None,
                 end_km=None,
                 *args,
                 **kwargs):

        super(RiverGraph, self).__init__(*args, **kwargs)

        #self.figure.set_facecolor('#bbbbbb')
        self.start_km = start_km
        self.end_km = end_km

        # Layout of the axes in the figure
        axes_left = 0.1
        axes_width = 0.87
        self.axes = self.figure.add_axes([axes_left,
                                          0.31,
                                          axes_width,
                                          0.49],
                                         ylabel='MHW overschrijding [m]')
        self.axes.grid(True, linestyle='-', color='lightgrey')
        self.axes.invert_xaxis()
        self.bar_axes = self.figure.add_axes([axes_left,
                                              0.10,
                                              axes_width,
                                              0.20],
                                             xlabel='Rivier Kilometer')
        self.bar_axes.grid(True, linestyle='-', color='lightgrey')
        self.bar_axes.set_ylim((-0.05, 1.1))
        self.bar_axes.get_yaxis().set_ticks([0.1, 0.3, 0.5, 0.7, 0.9])
        self.bar_axes.get_yaxis().set_ticklabels(['knelpt',
                                                  'Retent',
                                                  'Dijk',
                                                  'Onttr',
                                                  'Overig'])

        # bar style options and initializations
        self.legend_handles = []
        self.legend_labels = []
        self.polysize = 70
        self.circlesize = 40
        self.legend_markersize = 7
        self.legend_markeredgewidth = 1.2
        self.colormap = cm.cool
        self.patch_zorder = 10

    def png_response(self):

        # Do final tweaks after data has been added to the axes
        ylim_old = self.axes.get_ylim()
        ylim_new = (ylim_old[0],
                    ylim_old[1] + 0.15 * (ylim_old[1] - ylim_old[0]))
        self.axes.set_ylim(ylim_new)
        self.bar_axes.set_xlim(self.axes.get_xlim())
        for l in self.axes.get_xaxis().get_majorticklabels():
            l.set_visible(False)
        for l in self.bar_axes.get_yaxis().get_majorticklabels():
            l.set_horizontalalignment('left')
            l.set_position((-.08, 0))

        self.axes.set_autoscaley_on(False)
        self.axes.axhline(0,
                          color='#030303',
                          linestyle=':',
                          linewidth=3,
                          label='Nullijn')

        # Add fake patches at fake positions
        seed(0)
        xlim = self.axes.get_xlim()
        step = (max(xlim) - min(xlim)) / 40
        kms = arange(xlim[1] + step, xlim[0] - step, step)
        iters = 25

        self.add_diamonds([choice(kms) for i in range(iters)])
        self.add_uptriangles([choice(kms) for i in range(iters)])
        self.add_squares([choice(kms) for i in range(iters)])
        self.add_downtriangles([choice(kms) for i in range(iters)])
        self.add_circles([choice(kms) for i in range(iters)])
        self.bar_axes.text(0.5, 0.5, 'RANDOM DATA',
                           horizontalalignment='center',
                           verticalalignment='center',
                           transform=self.bar_axes.transAxes,
                           zorder=100,
                           size=50,
                           color='red',
                           alpha=0.2)

        # Add some test place_names
        self.add_text((xlim[1] + (xlim[0] - xlim[1]) * 0.2,
                       xlim[1] + (xlim[0] - xlim[1]) * 0.5,
                       xlim[1] + (xlim[0] - xlim[1]) * 0.7),
                      ('Plaats C',
                       'Plaats B',
                       'Plaats A'))

        self.legend()

        return super(RiverGraph, self).png_response()

    def add_diamonds(self, kms):
        """Add diamonds to bar_axes at specified kms.

        The self.legend_handles_labels receives the patch as well."""
        ypos = 0.9
        km_arr = array(kms).reshape(-1, 1)
        ypos_arr = array([ypos for x in kms]).reshape(-1, 1)
        offsets = concatenate((km_arr, ypos_arr), axis=1)

        collection = RegularPolyCollection(
            numsides=4,
            rotation=0,
            sizes=(self.polysize,),
            facecolors=self.colormap(ypos),
            offsets=offsets,
            transOffset=self.bar_axes.transData,
            zorder=self.patch_zorder)
        self.bar_axes.add_collection(collection)

        line = Line2D((0, 1),
                      (0, 0),
                      linestyle='',
                      marker='D',
                      markersize=self.legend_markersize,
                      markeredgewidth=self.legend_markeredgewidth,
                      markerfacecolor=self.colormap(ypos))
        label = 'Overig'
        self.legend_handles.append(line)
        self.legend_labels.append(label)

    def add_uptriangles(self, kms):
        """Add upward triangles to bar_axes at specified kms.

        The self.legend_handles_labels receives the patch as well."""
        ypos = 0.7
        km_arr = array(kms).reshape(-1, 1)
        ypos_arr = array([ypos for x in kms]).reshape(-1, 1)
        offsets = concatenate((km_arr, ypos_arr), axis=1)

        collection = RegularPolyCollection(
            numsides=3,
            rotation=0,
            sizes=(self.polysize,),
            facecolors=self.colormap(ypos),
            offsets=offsets,
            transOffset=self.bar_axes.transData,
            zorder=self.patch_zorder)
        self.bar_axes.add_collection(collection)

        line = Line2D((0, 1),
                      (0, 0),
                      linestyle='',
                      marker='^',
                      markersize=self.legend_markersize,
                      markeredgewidth=self.legend_markeredgewidth,
                      markerfacecolor=self.colormap(ypos))
        label = 'Groene rivieren / ontrekkingen'
        self.legend_handles.append(line)
        self.legend_labels.append(label)

    def add_squares(self, kms):
        """Add squares to bar_axes at specified kms.

        The self.legend_handles_labels receives the patch as well."""
        ypos = 0.5
        km_arr = array(kms).reshape(-1, 1)
        ypos_arr = array([ypos for x in kms]).reshape(-1, 1)
        offsets = concatenate((km_arr, ypos_arr), axis=1)

        collection = RegularPolyCollection(
            numsides=4,
            rotation=pi / 4,
            sizes=(self.polysize,),
            facecolors=self.colormap(ypos),
            offsets=offsets,
            transOffset=self.bar_axes.transData,
            zorder=self.patch_zorder)
        self.bar_axes.add_collection(collection)

        line = Line2D((0, 1),
                      (0, 0),
                      linestyle='',
                      marker='s',
                      markersize=self.legend_markersize,
                      markeredgewidth=self.legend_markeredgewidth,
                      markerfacecolor=self.colormap(ypos))
        label = 'Dijken / kades'
        self.legend_handles.append(line)
        self.legend_labels.append(label)

    def add_downtriangles(self, kms):
        """Add downward triangles to bar_axes at specified kms.

        The self.legend_handles_labels receives the patch as well."""
        ypos = 0.3
        km_arr = array(kms).reshape(-1, 1)
        ypos_arr = array([ypos for x in kms]).reshape(-1, 1)
        offsets = concatenate((km_arr, ypos_arr), axis=1)

        collection = RegularPolyCollection(
            numsides=3,
            rotation=pi,
            sizes=(self.polysize,),
            facecolors=self.colormap(ypos),
            offsets=offsets,
            transOffset=self.bar_axes.transData,
            zorder=self.patch_zorder)
        self.bar_axes.add_collection(collection)

        line = Line2D((0, 1),
                      (0, 0),
                      linestyle='',
                      marker='v',
                      markersize=self.legend_markersize,
                      markeredgewidth=self.legend_markeredgewidth,
                      markerfacecolor=self.colormap(ypos))
        label = 'Retentie'
        self.legend_handles.append(line)
        self.legend_labels.append(label)

    def add_circles(self, kms):
        """Add circles to bar_axes at specified kms.

        The self.legend_handles_labels receives the patch as well."""
        ypos = 0.1
        km_arr = array(kms).reshape(-1, 1)
        ypos_arr = array([ypos for x in kms]).reshape(-1, 1)
        offsets = concatenate((km_arr, ypos_arr), axis=1)

        collection = CircleCollection(
            sizes=(self.circlesize,),
            facecolors=self.colormap(ypos),
            offsets=offsets,
            transOffset=self.bar_axes.transData,
            zorder=self.patch_zorder)
        self.bar_axes.add_collection(collection)
        line = Line2D((0, 1),
                      (0, 0),
                      linestyle='',
                      marker='o',
                      markersize=self.legend_markersize,
                      markeredgewidth=self.legend_markeredgewidth,
                      markerfacecolor=self.colormap(ypos))
        label = 'Knelpunten'
        self.legend_handles.append(line)
        self.legend_labels.append(label)

    def add_text(self, kms, strs):
        ylim = self.axes.get_ylim()
        ypos = ylim[0] + (ylim[1] - ylim[0]) * 0.9
        for k, t in zip(kms, strs):
            self.axes.add_artist(Text(k,
                                      ypos,
                                      t,
                                      backgroundcolor='white',
                                      horizontalalignment='center'))
            self.axes.axvline(k,
                   color='#030303',
                   linestyle='--',
                   linewidth=1,
                   zorder=-10,
                   label='remove_from_legend')
            self.bar_axes.axvline(k,
                   color='#030303',
                   linestyle='--',
                   linewidth=1,
                   zorder=-10,
                   label=t)

    def legend(self, handles=None, labels=None):
        handles, labels = self.axes.get_legend_handles_labels()

        # Remove the labels from the vlines
        def f(tup):
            return tup[1] != 'remove_from_legend'
        handles, labels = zip(*filter(f, zip(handles, labels)))

        handles = list(handles)
        labels = list(labels)

        handles.extend(self.legend_handles)
        labels.extend(self.legend_labels)

        ncol = int(4 * self.width / 640)
        self.axes.legend(handles,
                         labels,
                         bbox_to_anchor=(0., 1.02, 1., .102),
                         loc=3,
                         ncol=ncol,
                         mode="expand",
                         borderaxespad=0.)
